sql/item.cc

```diff
@@ -2176,6 +2176,26 @@ Item::Type Item_name_const::type() const
 
 bool Item_name_const::fix_fields(THD *thd, Item **ref)
 {
+  /*
+    If we have either of the following:
+      ... WHERE foo=NAME_CONST(...)
+      ... JOIN ... ON foo=NAME_CONST(...)
+    then we have an opportunity to unwrap the NAME_CONST and
+    use the enclosed value directly, replacing NAME_CONST in
+    the parse tree with the value it encloses.
+   */
+  if ((thd->where == THD_WHERE::WHERE_CLAUSE ||
+       thd->where == THD_WHERE::ON_CLAUSE) &&
+    !value_item->fix_fields_if_needed(thd, &value_item))
+  {
+    thd->change_item_tree(ref, value_item);
+    if (value_item->collation.derivation != DERIVATION_NUMERIC)
+      value_item->collation.set(value_item->collation.collation,
+                                DERIVATION_IMPLICIT);
+    return FALSE;
```

Comment:
The first and obvious question: if we reach this return statement, then
```
    base_flags|= item_base_t::FIXED;
```
is not done for the Item_const object, and item remains unfixed.